### ISSUE TITLE
Spamming of Logs with Warning: A reader instance was requested via setReadOnly...

### ISSUE THREAD

**Issue by datenbrille** [awslabs/aws-advanced-jdbc-wrapper#549]
### Describe the bug

This behaviour/bug is related to https://github.com/awslabs/aws-advanced-jdbc-wrapper/issues/547 and a mitigation we have implemented to ensure our production systems can still use the Read/Write split. As a side effect with the custom Read/Write Plugin we had to switch the logging to SLF4J/Logback (We did not had the bridge in place before the OOM mitigation). 

Since then our logs get spammed with the warning and I wonder why this is raised on our Aurora PostgreSQL 2 node cluster.

### Expected Behavior

The system is able to resolve in a stable way the host list of an aurora cluster.

### What plugins are used? What other connection properties were set?

AwsSecretsManagerConnectionPluginFactory, CustomReadWriteSplittingPluginFactory,FailoverConnectionPluginFactory,HostMonitoringConnectionPluginFactory

### Current Behavior

After putting the system into production and on to load (around 250req/sec) the driver starts to spam the logs with a rate of more then 60 messages per second.

### Reproduction Steps

Having a system which is frequently switching on the same machine/instance between read and write connections all the time.

### Possible Solution

* Change the log level to debug/ trace
* Do not log and present a metric instead

### The AWS Advanced JDBC Driver version used

2.2.1, 2.2.2

### JDK version used

AWS Corretto 17.0.7

### Operating System and version

Amazon Linux on AWS Elastic Beanstalk

**Comment by datenbrille** [comment]
After some further investigations the behaviour is related to our Custom changes, because we are eagerly releasing now the reader connections back to the pool and with that we hitting the `initializeReaderConnection` way more often then before. I debug further and I came to the conclusion that the Topology query is retuning sometimes only one node in our 2 node cluster. If this happens the code will start to throw the warning.

**Comment by crystall-bitquill** [comment]
Hi @datenbrille,

Thanks for reaching out and raising this issue.

We'll take a look at this and keep you updated as we investigate.

Thank you for your patience!

**Comment by aaronchung-bitquill** [comment]
HI @datenbrille 

I am currently trying to reproduce this issue. In the mean time, could you confirm that each instance in your DB cluster is healthy? 

The topology query filters out entries that are stale (entries whose last updated timestamps are > 5minutes). Stale entries can sometimes be the result of unhealthy hosts/instances. 

Also, would you be able to share information about the following?

- what frameworks/libraries are you using? (ex. Spring? Hikari?)
- if possible to share, what changes/modifications/mitigations did you make to the CustomReadWriteSplittingPlugin? 

Thank you.

**Comment by datenbrille** [comment]
Hi @aaronchung-bitquill,

we are using Aurora PosgreSQL 14.7 with a 2 node cluster of 1 reader and 1 writer. The custom Plugin is a copy of the R/W Split one with the modifications in the PR of https://github.com/awslabs/aws-advanced-jdbc-wrapper/pull/548. With the change we intialize the connection more than usual and hit the warn log to frequently so to say.

We are using HikariCP together with a Spring 5.3 (no Boot) stack.

I am currently trying to isolate the problem in a dedicated application with our prod. setup to see where we loose the one node. What I can see so fare it only happens if we create a lot of new connections to the DB. In single thread/connection mode it does not happen.

Best Karl

**Comment by aaronchung-bitquill** [comment]
Hi @datenbrille

Wanted to give you a bit of an update. There is a fix resulting from  [#547](https://github.com/awslabs/aws-advanced-jdbc-wrapper/issues/547) that is available in the latest [snapshot build](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/UsingTheJdbcDriver.md#using-a-snapshot-of-the-driver). 
Can you check it out using the regular ReadWriteSplittingPlugin and let us know if that works? 

Thank you!

**Comment by datenbrille** [comment]
Hi @aaronchung-bitquill,

I tested the release and the problem is still there and I found the source but not the why. If you do the topology query on a reader connection the resultset in case of 2 node cluster is 1 and only contains only the writer. I am able to reproduce this in our setup if I increase the load on the system and with that the load of the two hikari pools (1 main and 1 for the reader connection) it happens that the connection to resolve the topology is actually a reader connection.

Does it make sense to you?

Maybe you can give me a hint where this can happen. It looks like the connection leaks somewhere.

Best Karl

**Comment by datenbrille** [comment]
Hi @aaronchung-bitquill,

I did some further digging and I believe this is a timing issue in the R/W split Plugin. If the current connection is switched in

```java
private void switchCurrentConnectionTo(final Connection newConnection, final HostSpec newConnectionHost)
            throws SQLException {
        final Connection currentConnection = this.pluginService.getCurrentConnection();
        if (currentConnection == newConnection) {
            return;
        }

        transferSessionStateOnReadWriteSplit(newConnection);
        // new connection is a reader
        this.pluginService.setCurrentConnection(newConnection, newConnectionHost);
        log.trace(Messages.get(
                "ReadWriteSplittingPlugin.settingCurrentConnection", new Object[] {newConnectionHost.getUrl()}));
    }
```
and we hit the Topology refresh method in the AuroraHostListProvider with no Connection given in the this method

```java
@Override
  public List<HostSpec> refresh() throws SQLException {
    return this.refresh(null);
  }
```

We take the current Connection from the PluginServiceImpl. This is currently then a reader connection and with that only one node is resolved.

Does this help to track down the Bug/Behaviour. I think the topology should not be resolved from any reader connection by either ignoring the topology refresh on a reader connection or always provide a cluster/writer connection for the refresh methods.

Best Karl

**Comment by datenbrille** [comment]
I am happy to create a PR if you point me into a direction of a preferred the solution.

**Comment by aaronchung-bitquill** [comment]
Hi @datenbrille 

Thank you for diving deep, sharing the detailings all of your findings, and even coming up with ideas for potential solutions.

The main concern I have is why the topology query using reader connections is incorrectly returning a topology of size = 1 instead of 2. If the connection is a reader, the topology query should still get the right topology. I am afraid that  avoiding using the reader connection for the topology query is a work around that does not address underlying issue, and instead masks it.

I'll try to reproduce the issue using two Hikari pools like you mentioned most recently. 

Also, did the latest version or build help alleviate any of this issue on your production systems?

Thank you

**Comment by datenbrille** [comment]
@aaronchung-bitquill, the last release fixed the OOM and the memory leak in the reader pool yes. But not the log spamming. We fixed that by having still our custom plugin and set the log level to off.

Regarding the topology query: This happens too if you do not have a second hikari pool for the reader. It will happen if you have R/W Split on and only use reader connections with Transaction level read-only and the topology refresh happens if the the current connection is reader connection.

So you test should only use reader connections and reduce the topo cache update to 10 seconds which should increase the probability.

**Comment by aaronchung-bitquill** [comment]
Hi @datenbrille 

I wanted to check in and give you an update. 

We were able to reproduce and find the root cause. The root cause is that in some versions of Aurora Postgres, the `last_update_timestamp` field from the topology query can be null. This is a scenario that the driver currently does not account for. The driver uses this field to filter out stale entries, and entries with a null `last_update_timestamp` were being filtered out as well. 

However, a code change in the driver to address this is in the works. I'll let you know once the code is merged in and ready for testing. 

Thank you!

**Comment by aaronchung-bitquill** [comment]
Hi @datenbrille 

We've merged in a fix. Could you kindly check out our [snapshot](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/UsingTheJdbcDriver.md#using-a-snapshot-of-the-driver) build and let us know if the issue persists?

Thank you!

**Comment by aaronchung-bitquill** [comment]
Thank you @datenbrille!

We are currently preparing for release that includes this fix. We'll let you know when the release is available.

**Comment by crystall-bitquill** [comment]
Hi @datenbrille,

Version [2.2.4](https://search.maven.org/artifact/software.amazon.jdbc/aws-advanced-jdbc-wrapper/2.2.4/jar?eh=) of the AWS JDBC Driver is now available, and it includes the fix for this issue. This ticket will be closed, but please feel free to reopen it or to create a new issue if you have other questions or concerns.

Thank you!
